### ISSUE TITLE
BUG: scipy.stats.kstest returns NaN starting in scipy 1.12

### ISSUE THREAD

**Issue by jakevdp** [scipy/scipy#20386]
### Describe your issue.

When the input contains both negative and positive infinity, `scipy.stats.kstest` returns `NaN`. This appears to have been introduced in scipy 1.12.0

### Reproducing Code Example

```python
import numpy as np
import scipy

vals = [-np.inf, 0, 1, np.inf]
result = scipy.stats.kstest(vals, scipy.stats.cauchy.cdf)
print(f"scipy v{scipy.__version__}: {result}")
```

Output in scipy 1.11.4:
```
scipy v1.11.4: KstestResult(statistic=0.25, pvalue=0.90625, statistic_location=0.0, statistic_sign=-1)
```
Output in scipy 1.12.0:
```
scipy v1.12.0: KstestResult(statistic=nan, pvalue=nan, statistic_location=nan, statistic_sign=nan)
```
Output in scipy 1.13.0:
```
scipy v1.13.0: KstestResult(statistic=nan, pvalue=nan, statistic_location=nan, statistic_sign=nan)
```

### SciPy/NumPy/Python version and system information

```shell
>>> import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info); scipy.show_config()
1.13.0 1.25.2 sys.version_info(major=3, minor=10, micro=12, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /usr/local/include
    lib directory: /usr/local/lib
    name: openblas
    openblas configuration: USE_64BITINT=0 DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= ZEN MAX_THREADS=64
    pc file directory: /usr/local/lib/pkgconfig
    version: 0.3.26.dev
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /usr/local/include
    lib directory: /usr/local/lib
    name: openblas
    openblas configuration: USE_64BITINT=0 DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= ZEN MAX_THREADS=64
    pc file directory: /usr/local/lib/pkgconfig
    version: 0.3.26.dev
  pybind11:
    detection method: config-tool
    include directory: unknown
    name: pybind11
    version: 2.12.0
Compilers:
  c:
    commands: cc
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  c++:
    commands: c++
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.0.10
  fortran:
    commands: gfortran
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  pythran:
    include directory: ../../tmp/pip-build-env-0blqy1or/overlay/lib/python3.10/site-packages/pythran
    version: 0.15.0
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
Python Information:
  path: /opt/python/cp310-cp310/bin/python
  version: '3.10'
```

**Comment by mdhaber** [comment]
As a quick fix, you can use:
```python3
stats.kstest(vals, scipy.stats.cauchy.cdf, _no_deco=True)
# KstestResult(statistic=0.25, pvalue=0.90625, statistic_location=0.0, statistic_sign=-1)
```

**Comment by mdhaber** [comment]
I ran into the same thing recently. 

What was new in SciPy 1.12 is that `kstest` got `nan_policy` support, and that uses the old `_contains_nan` function to check whether there are NaNs. 

The `_contains_nan` function has (for a long time) detected NaNs by summing all elements of the array and checking whether the result is NaN. The rationale was that this saves memory, but when you add `-inf + inf`, a NaN is introduced. 

I'd suggest we get rid of this "summation" method and just do the obvious thing. `contains_nan = np.isnan(a).any()`. Er, rather, I would make the change in gh-20292, which adds array API support to `_contains_nan` and `stats.moment`, so it would be `contains_nan = xp.any(xp.isnan(a))`.

**Comment by thomasjpfan** [comment]
scikit-learn has a similar `_assert_all_finite` where it does a sum check first (for speed). If the sum check fails then it does a `xp.any(xp.isnan(...))` to see if there was an actual nan.

XREF: https://github.com/scikit-learn/scikit-learn/blob/f7026e575a494b47b50557932b5c3ce0688fdc72/sklearn/utils/validation.py#L121-L126

**Comment by mdhaber** [comment]
Thanks. I guess I'm not seeing the advantage in our case, though. For instance, as soon as the array gets into `ks_1samp`, [two arrays](https://github.com/scipy/scipy/blob/7dcd8c59933524986923cde8e9126f5fc2e6b30b/scipy/stats/_stats_py.py#L8371) of the same size are created (and many intermediate arrays of the same size are created within the call to `cdf`). One of these is the same dtype as the original array, and the other is always `float64`. The temporary boolean array seems insignificant in comparison, so I'd think overall it would be worth the make the code a bit simpler and faster?

**Comment by fancidev** [comment]
A possible trick is to use `xp.max` instead of `xp.sum`.

**Comment by thomasjpfan** [comment]
In scikit-learn, we benchmarked our inf/nan check for 2d arrays and found that the memory pressure was much lower for valid input with no nans and infs.

For SciPy, if it is all 1D arrays, then it should be okay.

**Comment by fancidev** [comment]
In addition, `numpy.sum` performs pairwise summation, which is an overkill for detecting `nan` … It would be best if the Array API had a `count` function, so that you could do `count(x,nan)>0`.

**Comment by lucascolley** [comment]
Will we want this in the next bug fix release, or is it okay to wait for the next minor release?
